src/python/pants/backend/python/dependency_inference/module_mapper.py

```diff
@@ -308,6 +311,24 @@ def providers_for_module(
         )
 
 
+@functools.cache
+def generate_mappings_from_pattern(proj_name: str, fallback: str) -> Iterable[str]:
+    """Generate a list of possible module mappings from a project name using a regex pattern.
+
+    e.g. google-cloud-foo -> [google.cloud.foo, google.cloud.foo_v1, google.cloud.foo_v1beta1]
+    Should eliminate the need to "manually" add a mapping for every service
+    proj_name: The project name to generate mappings for e.g google-cloud-datastream
+    """
+    pattern_values = []
+    for match_pattern, replace_patterns in DEFAULT_MODULE_PATTERN_MAPPING.items():
+        if match_pattern.match(proj_name) is not None:
+            pattern_values = [
+                match_pattern.sub(replace_pattern, proj_name)
+                for replace_pattern in replace_patterns
+            ]
+    return pattern_values + [fallback]
```

Comment:
**question:** should the fallback be appended?
i initially started appending the fallback, which is almost always just the module with `_` e.g. `google-cloud-datastream` -> `google_cloud_datastream`. but, that is likely useless if we find a match. so perhaps this should be `return pattern_values or [fallback]` so we only use the `fallback` if no other pattern matched worked.

Reply:
done. combined with your other recommendation below.